File: plugin/zeppelin/src/main/scala/jetbrains/zeppelin/idea/toolwindow/ZeppelinToolWindowFactory.scala
package jetbrains.zeppelin.idea.toolwindow

import com.intellij.openapi.actionSystem.{ActionManager, DefaultActionGroup}
import com.intellij.openapi.fileEditor._
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.SimpleToolWindowPanel
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.wm.{ToolWindow, ToolWindowFactory}
import com.intellij.ui.content.{Content, ContentFactory}
import jetbrains.zeppelin.components.ZeppelinComponent
import jetbrains.zeppelin.idea.toolwindow.actions._
import jetbrains.zeppelin.utils.ZeppelinLogger

/**
  * Factory that creates a Zeppelin tool window
  */
class ZeppelinToolWindowFactory extends ToolWindowFactory {

  override def createToolWindowContent(project: Project, toolWindow: ToolWindow): Unit = {
    toolWindow.getContentManager.addContent(createLogPanel(project))
    toolWindow.getContentManager.addContent(createInterpretersPanel(project))

    addAutoUpdate(project)
  }

  override def init(toolWindow: ToolWindow): Unit = {
    toolWindow.setStripeTitle("Zeppelin")
  }

  private def addAutoUpdate(project: Project): Unit = {
    ZeppelinComponent.connectionFor(project).updateInterpreterList(force = true)

    val connection = project.getMessageBus.connect(project)
    connection
      .subscribe(FileEditorManagerListener.FILE_EDITOR_MANAGER, new FileEditorManagerListener() {
        override def selectionChanged(event: FileEditorManagerEvent): Unit = {
          ZeppelinComponent.connectionFor(project).updateInterpreterList()
        }
      })
  }

  private def createInterpretersPanel(project: Project): Content = {
    val panel = new SimpleToolWindowPanel(false, true)

    val interpretersView = ZeppelinComponent.connectionFor(project).interpretersView
    panel.setContent(interpretersView)
    val toolbar = createInterpretersToolbar(project, interpretersView)
    panel.setToolbar(toolbar.getComponent)
    val content = ContentFactory.SERVICE.getInstance.createContent(panel, "Interpreters", true)
    Disposer.register(project, interpretersView)
    content
  }

  private def createInterpretersToolbar(project: Project, interpreters: InterpretersView) = {
    val group = new DefaultActionGroup
    group.add(new RefreshInterpretersAction())
    val toolbar = ActionManager.getInstance.createActionToolbar("left", group, false)
    toolbar.setTargetComponent(interpreters.getComponent)
    toolbar
  }

  private def createLogPanel(project: Project): Content = {
    val panel = new SimpleToolWindowPanel(false, true)
    val console = new ZeppelinConsole(project)
    ZeppelinLogger.initOutput(console)
    panel.setContent(console)
    val toolbar = createLogToolbar(project, console)
    panel.setToolbar(toolbar.getComponent)
    val content = ContentFactory.SERVICE.getInstance.createContent(panel, "Log", true)
    Disposer.register(project, console)
    content
  }

  private def createLogToolbar(project: Project, console: ZeppelinConsole) = {
    val group = new DefaultActionGroup
    group.add(new RefreshInterpretersAction())
    group.add(new RunCodeAction(project))
    group.addSeparator()
    group.add(new OpenGlobalSettingsFormAction())
    group.addSeparator()
    group.add(new ClearLogActionConsole(console))
    val toolbar = ActionManager.getInstance.createActionToolbar("left", group, false)
    toolbar.setTargetComponent(console.getComponent)
    toolbar
  }
}

object ZeppelinToolWindowFactory {
  val ID = "zeppelin-shell-toolwindow"
}

File: zeppelin/src/org/intellij/plugin/zeppelin/idea/toolwindow/ZeppelinToolWindowFactory.kt
package org.intellij.plugin.zeppelin.idea.toolwindow

import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.ActionToolbar
import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.fileEditor.FileEditorManagerEvent
import com.intellij.openapi.fileEditor.FileEditorManagerListener
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.SimpleToolWindowPanel
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.wm.ToolWindow
import com.intellij.openapi.wm.ToolWindowFactory
import com.intellij.ui.content.Content
import com.intellij.ui.content.ContentFactory
import com.intellij.util.messages.MessageBusConnection
import org.intellij.plugin.zeppelin.components.ZeppelinComponent
import org.intellij.plugin.zeppelin.idea.toolwindow.actions.ClearLogActionConsole
import org.intellij.plugin.zeppelin.idea.toolwindow.actions.OpenGlobalSettingsFormAction
import org.intellij.plugin.zeppelin.idea.toolwindow.actions.RefreshInterpretersAction
import org.intellij.plugin.zeppelin.idea.toolwindow.actions.RunCodeAction
import org.intellij.plugin.zeppelin.utils.ZeppelinLogger

/**
 * Factory that creates a Zeppelin tool window
 */
class ZeppelinToolWindowFactory : ToolWindowFactory {
    override fun createToolWindowContent(project: Project, toolWindow: ToolWindow) {
        toolWindow.contentManager.addContent(createLogPanel(project))
        toolWindow.contentManager.addContent(createInterpretersPanel(project))
        addAutoUpdate(project)
    }

    override fun init(toolWindow: ToolWindow) {
        toolWindow.stripeTitle = "Zeppelin"
    }

    private fun addAutoUpdate(project: Project) {
        ZeppelinComponent.connectionFor(project).updateInterpreterList(true)
        val connection: MessageBusConnection = project.messageBus.connect(project)
        connection.subscribe(FileEditorManagerListener.FILE_EDITOR_MANAGER, object : FileEditorManagerListener {
            override fun selectionChanged(event: FileEditorManagerEvent) = ZeppelinComponent.connectionFor(project).updateInterpreterList(true)
        })
    }

    private fun createInterpretersPanel(project: Project): Content {
        val panel = SimpleToolWindowPanel(false, true)
        val interpretersView: InterpretersView = ZeppelinComponent.connectionFor(project).interpretersView
        panel.setContent(interpretersView)
        val toolbar: ActionToolbar = createInterpretersToolbar(interpretersView)
        panel.setToolbar(toolbar.component)
        val content: Content = ContentFactory.SERVICE.getInstance().createContent(panel, "Interpreters", true)
        Disposer.register(project, interpretersView)
        return content
    }

    private fun createInterpretersToolbar(interpreters: InterpretersView): ActionToolbar {
        val group = DefaultActionGroup()
        group.add(RefreshInterpretersAction())
        val toolbar: ActionToolbar = ActionManager.getInstance().createActionToolbar("left", group, false)
        toolbar.setTargetComponent(interpreters)
        return toolbar
    }

    private fun createLogPanel(project: Project): Content {
        val panel = SimpleToolWindowPanel(false, true)
        val console = ZeppelinConsole(project)
        ZeppelinLogger.output = console
        panel.setContent(console)
        val toolbar: ActionToolbar = createLogToolbar(console)
        panel.setToolbar(toolbar.component)
        val content: Content = ContentFactory.SERVICE.getInstance().createContent(panel, "Log", true)
        Disposer.register(project, console)
        return content
    }

    private fun createLogToolbar(console: ZeppelinConsole): ActionToolbar {
        val group = DefaultActionGroup()
        group.add(RefreshInterpretersAction())
        group.add(RunCodeAction())
        group.addSeparator()
        group.add(OpenGlobalSettingsFormAction())
        group.addSeparator()
        group.add(ClearLogActionConsole(console))
        val toolbar: ActionToolbar = ActionManager.getInstance().createActionToolbar("left", group, false)
        toolbar.setTargetComponent(console.component)
        return toolbar
    }

    companion object {
        const val ID: String = "zeppelin-shell-toolwindow"
    }
}